### PR TITLE
adds response header forwarding for providers

### PR DESCRIPTION
**Pull request by akshaydeo**
## Summary

Adds streaming improvements by reducing accumulation to o(n) vs o(n2) loops

## Type of change

- [ ] Bug fix
- [ ] Feature
- [x] Refactor
- [ ] Documentation
- [ ] Chore/CI

## Affected areas

- [x] Core (Go)
- [ ] Transports (HTTP)
- [ ] Providers/Integrations
- [ ] Plugins
- [ ] UI (Next.js)
- [ ] Docs

## How to test

Describe the steps to validate this change. Include commands and expected outcomes.

```sh
# Core/Transports
go version
go test ./...

# UI
cd ui
pnpm i || npm i
pnpm test || npm test
pnpm build || npm run build
```

If adding new configs or environment variables, document them here.

## Screenshots/Recordings

If UI changes, add before/after screenshots or short clips.

## Breaking changes

- [ ] Yes
- [x] No

If yes, describe impact and migration instructions.

## Related issues

Link related issues and discussions. Example: Closes #123

## Security considerations

Note any security implications (auth, secrets, PII, sandboxing, etc.).

## Checklist

- [x] I read `docs/contributing/README.md` and followed the guidelines
- [x] I added/updated tests where appropriate
- [x] I updated documentation where needed
- [x] I verified builds succeed (Go and UI)
- [x] I verified the CI pipeline passes locally if applicable

### CODE REVIEW

**Comment by akshaydeo** [comment]
* **#1868** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/1868?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/maximhq/bifrost/1868?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: core/schemas/provider.go [hunk 1]
```diff
@@ -14,7 +14,7 @@ const (
 	DefaultRequestTimeoutInSeconds = 30
 	DefaultBufferSize              = 5000
 	DefaultConcurrency             = 1000
-	DefaultStreamBufferSize        = 5000
+	DefaultStreamBufferSize        = 256
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/schemas/provider.go:17]: This is intentional. The reduction from 5000 to 256 was part of the streaming performance improvements in this PR. A buffer of 5000 was significantly over-provisioned — most streaming responses have far fewer chunks in-flight at once, and the large buffer was causing unnecessary memory overhead per stream. The new value of 256 still provides ample buffering while reducing memory usage. This was validated with the streaming performance test included in the same changeset.

File: plugins/logging/main.go [hunk 2]
```diff
@@ -511,340 +499,142 @@
...
-
-		if result != nil {
-			logMsg.Latency = result.GetExtraFields().Latency
-		} else {
-			logMsg.Latency = 0
-		}
+	// Retrieve pending input data from PreLLMHook
+	pendingVal, hasPending := p.pendingLogs.LoadAndDelete(requestID)
+	if !hasPending {
+		p.logger.Warn("no pending log data found for request %s, skipping log write", requestID)
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply plugins/logging/main.go]: False positive — `requestID` is a request correlation ID (sourced from `schemas.BifrostContextKeyRequestID`), not an API key or secret. It is intentionally logged for request tracing and debugging purposes. The CodeQL data flow analysis incorrectly associates this variable with sensitive data.

File: framework/logstore/rdb.go [hunk 3]
```diff
@@ -946,6 +958,28 @@ func (s *RDBLogStore) FindAll(ctx context.Context, query any, fields ...string)
...
+func (s *RDBLogStore) FindAllDistinct(ctx context.Context, query any, fields ...string) ([]*Log, error) {
+	var logs []*Log
+	db := s.db.WithContext(ctx).Where(query)
+	if len(fields) > 0 {
+		args := make([]interface{}, len(fields))
+		for i, f := range fields {
+			args[i] = f
+		}
+		db = db.Distinct(args...)
+	}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/logstore/rdb.go:1120]: Fixed — added an `allowedDistinctLogColumns` allowlist that validates each field before passing to GORM's `Distinct()`. Invalid fields now return an error. This is solved, can you check and resolve if done properly?

File: core/providers/replicate/replicate.go [hunk 4]
```diff
@@ -127,28 +127,31 @@ func createPrediction(
...
 
+	// Extract provider response headers before releasing the response
+	providerResponseHeaders := providerUtils.ExtractProviderResponseHeaders(resp)
+
 	// Parse response
 	body, decodeErr := providerUtils.CheckAndDecodeBody(resp)
 	if decodeErr != nil {
-		return nil, nil, latency, providerUtils.NewBifrostOperationError(schemas.ErrProviderResponseDecode, decodeErr, schemas.Replicate)
+		return nil, nil, latency, nil, providerUtils.NewBifrostOperationError(schemas.ErrProviderResponseDecode, decodeErr, schemas.Replicate)
 	}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/replicate/replicate.go:146]: Fixed — moved `ExtractProviderResponseHeaders` before the status code check in both `createPrediction` and `getPrediction`. Provider response headers are now returned on all error paths (non-2xx, decode errors, handler errors). This is solved, can you check and resolve if done properly?

File: plugins/logging/writer.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+		if entry.callback != nil {
+			callbacks = append(callbacks, cbPair{cb: entry.callback, log: entry.log})
+		}
+	}
+	if len(callbacks) > 0 {
+		go func() {
+			for _, pair := range callbacks {
+				pair.cb(pair.log)
+			}
+		}()
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply plugins/logging/writer.go]: Fixed — added `defer recover()` with panic logging to the async callback goroutine. A panicking callback will now be caught and logged instead of crashing the process. This is solved, can you check and resolve if done properly?

File: ui/app/pprof/page.tsx [hunk 6]
```diff
@@ -0,0 +1,764 @@
...
+				<button
+					type="button"
+					onClick={(e) => {
+						e.stopPropagation();
+						const filePath = getStackFilePath(group.stack);
+						if (filePath) onSkip(filePath);
+					}}
+					data-testid="pprof-goroutine-skip"
+					className="shrink-0 rounded p-1.5 text-zinc-600 opacity-0 transition-opacity group-hover:opacity-100 hover:bg-zinc-700 hover:text-zinc-300"
+					title="Hide goroutines from this file"
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply ui/app/pprof/page.tsx:260]: Fixed — added `focus-visible:opacity-100`, `group-focus-within:opacity-100`, and `focus-visible:ring-2 focus-visible:ring-zinc-500` classes so the skip button is visible and has a focus ring for keyboard users. This is solved, can you check and resolve if done properly?

File: core/providers/openai/openai.go [hunk 7]
```diff
@@ -1564,6 +1585,7 @@ func HandleOpenAIResponsesStreaming(
 
 			// Parse into bifrost response
 			var response schemas.BifrostResponsesStreamResponse
+			// TODO fix this
 			if customResponseHandler != nil {
 				rawRequest, rawResponse, bifrostErr := customResponseHandler([]byte(jsonData), &response, nil, false, false)
 				if bifrostErr != nil {
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/openai/openai.go:1597]: This is a known TODO (line 1588: `// TODO fix this`) — the `customResponseHandler` path in `HandleOpenAIResponsesStreaming` is not yet fully wired up. This will be addressed in a dedicated follow-up once the Responses API streaming handler is complete.

File: plugins/logging/main.go [hunk 8]
```diff
@@ -511,340 +499,148 @@ func (p *LoggerPlugin) PostLLMHook(ctx *schemas.BifrostContext, result *schemas.
...
+		return result, bifrostErr, nil
+	}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply plugins/logging/main.go:542]: This early return is pre-existing behavior — not introduced by this PR. The scenario (PostLLMHook firing without a matching PreLLMHook pending entry) is a rare edge case. Building a best-effort fallback is a valid improvement but better suited for a separate PR to properly test the logging pipeline behavior change. Tracked for follow-up.

File: core/providers/vertex/vertex.go [hunk 9]
```diff
@@ -523,6 +523,7 @@ func (provider *VertexProvider) ChatCompletion(ctx *schemas.BifrostContext, key
 		if request.Model != deployment {
 			response.ExtraFields.ModelDeployment = deployment
 		}
+		response.ExtraFields.ProviderResponseHeaders = providerUtils.ExtractProviderResponseHeaders(resp)
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/vertex/vertex.go:529]: Fixed — added `ProviderResponseHeaders` extraction in all 5 missing Vertex methods: Rerank, VideoGeneration, VideoRetrieve, VideoDownload (via scoped variable for the URL download branch), and CountTokens (both Anthropic and Gemini branches). See updated code.

File: framework/streaming/chat.go [hunk 10]
```diff
@@ -477,18 +482,13 @@ func (a *Accumulator) processChatStreamingResponse(ctx *schemas.BifrostContext,
...
+	// Non-final chunk: skip expensive rebuild since no consumer uses intermediate data.
+	// Both logging and maxim plugins return early when !isFinalChunk.
 	return &ProcessedStreamResponse{
 		RequestID:  requestID,
 		StreamType: streamType,
 		Provider:   provider,
 		Model:      model,
-		Data:       data,
+		Data:       nil,
 	}, nil
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/streaming/chat.go:493]: Fixed — added nil guards for `Data` at the top of both `ToBifrostResponse()` (returns nil) and `applyStreamingOutputToEntry()` (early return). All current callers already guard with `isFinalChunk`, so this is a defensive safety net. See updated code.

File: framework/tracing/tracer.go [hunk 11]
```diff
@@ -226,256 +226,23 @@ func (t *Tracer) GetDeferredSpanID(traceID string) string {
...
-				Message: msg,
-			},
-		}
-		result.ChatResponse.Choices = append(result.ChatResponse.Choices, choice)
-	}
-
-	return result
+	ttftMs, chunkCount := t.store.GetAccumulatedData(traceID)
+	return nil, ttftMs, chunkCount
 }
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/tracing/tracer.go:246]: Fixed — updated the interface doc for `GetAccumulatedChunks` in `core/schemas/tracer.go` to clarify that the response is always nil and only TTFT + chunk count are meaningful. See updated code.

**Comment by akshaydeo** [comment]
@coderabbitai full review

File: transports/bifrost-http/handlers/inference.go [hunk 12]
```diff
@@ -740,6 +747,7 @@ func (h *CompletionHandler) listModels(ctx *fasthttp.RequestCtx) {
 		}
 	}
 
+	forwardProviderHeaders(ctx, resp.ExtraFields.ProviderResponseHeaders)
 	// Send successful response
 	SendJSON(ctx, resp)
 }
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/handlers/inference.go]: This is intentional. Every callsite follows the pattern:

```go
resp, bifrostErr := h.client.SomeRequest(...)
if bifrostErr != nil {
    SendBifrostError(ctx, bifrostErr)
    return
}
forwardProviderHeaders(ctx, resp.ExtraFields.ProviderResponseHeaders)
```

`resp` is guaranteed non-nil by the error-check-early-return contract — all Bifrost provider methods return either `(resp, nil)` or `(nil, err)`, never `(nil, nil)`. Additionally, `ExtraFields` is a value type (`BifrostResponseExtraFields`), not a pointer, so it cannot be nil when `resp` is non-nil.

Adding nil guards at 30+ callsites would be defensive noise with no real protection.

**Comment by akshaydeo** [comment]
@coderabbitai full review

File: core/providers/replicate/replicate.go [hunk 13]
```diff
@@ -3073,7 +3100,9 @@ func (provider *ReplicateProvider) FileUpload(ctx *schemas.BifrostContext, key s
 		return nil, bifrostErr
 	}
 
-	return replicateResp.ToBifrostFileUploadResponse(providerName, latency, sendBackRawRequest, sendBackRawResponse, rawRequest, rawResponse), nil
+	fileResponse := replicateResp.ToBifrostFileUploadResponse(providerName, latency, sendBackRawRequest, sendBackRawResponse, rawRequest, rawResponse)
+	fileResponse.ExtraFields.ProviderResponseHeaders = providerUtils.ExtractProviderResponseHeaders(resp)
+	return fileResponse, nil
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/replicate/replicate.go:3134]: Valid observation. `BifrostErrorExtraFields` doesn't currently have a `ProviderResponseHeaders` field, so error responses from File API methods can't carry provider headers without a schema change. This is tracked as a follow-up enhancement. Note that the core non-streaming and streaming paths in this PR have been fixed to forward headers on non-200 responses.

File: plugins/logging/operations.go [hunk 14]
```diff
@@ -434,13 +434,203 @@ func (p *LoggerPlugin) updateStreamingLogEntry(
...
+			entry.ErrorDetails = string(data)
+		}
+		latF := float64(streamResponse.Data.Latency)
+		entry.Latency = &latF
+		return
+	}
+
+	entry.Status = "success"
+	latF := float64(streamResponse.Data.Latency)
+	entry.Latency = &latF
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply plugins/logging/operations.go:477]: This is intentional. `applyStreamingOutputToEntry` is only ever called when `isFinalChunk` is true (see `main.go:587`), so setting `entry.Status = "success"` is correct — it always represents the finalized stream result. The function also handles error cases first (checking `streamResponse.Data.ErrorDetails != nil` before setting success status).

File: core/providers/cohere/cohere.go [hunk 15]
```diff
@@ -173,24 +173,27 @@ func (provider *CohereProvider) completeRequest(ctx *schemas.BifrostContext, jso
...
 
 	// Handle error response
 	if resp.StatusCode() != fasthttp.StatusOK {
-		return nil, latency, parseCohereError(resp, meta)
+		return nil, latency, nil, parseCohereError(resp, meta)
 	}
 
+	// Extract provider response headers before releasing the response
+	providerResponseHeaders := providerUtils.ExtractProviderResponseHeaders(resp)
+
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/cohere/cohere.go]: Fixed — moved `ExtractProviderResponseHeaders` before the status code check across all providers (Cohere, Anthropic, Gemini, OpenAI, Perplexity, HuggingFace, Bedrock, Azure, Elevenlabs, Replicate) so both streaming and non-streaming error responses now forward provider headers (rate-limit, request-id, etc.). This is solved, can you check and resolve if done properly?

File: core/providers/gemini/gemini.go [hunk 16]
```diff
@@ -1531,21 +1550,16 @@ func (provider *GeminiProvider) TranscriptionStream(ctx *schemas.BifrostContext,
...
+				errorStr, _ := errorNode.Raw()
 				bifrostErr := &schemas.BifrostError{
 					Type:           schemas.Ptr("gemini_api_error"),
 					IsBifrostError: false,
 					Error: &schemas.ErrorField{
-						Message: fmt.Sprintf("Gemini API error: %v", errorCheck["error"]),
-						Error:   fmt.Errorf("stream error: %v", errorCheck["error"]),
+						Message: fmt.Sprintf("Gemini API error: %s", errorStr),
+						Error:   fmt.Errorf("stream error: %s", errorStr),
 					},
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/gemini/gemini.go:1567]: Fixed — changed to lowercase `"gemini api error: %s"` per Go conventions. This is solved, can you check and resolve if done properly?

File: core/providers/openai/openai.go [hunk 17]
```diff
@@ -176,6 +176,7 @@ func ListModelsByKey(
 	response.ExtraFields.Provider = providerName
 	response.ExtraFields.RequestType = schemas.ListModelsRequest
 	response.ExtraFields.Latency = latency.Milliseconds()
+	response.ExtraFields.ProviderResponseHeaders = providerUtils.ExtractProviderResponseHeaders(resp)
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/openai/openai.go]: Fixed — added `"authorization"` and `"cookie"` to the `providerResponseFilterHeaders` map in `core/providers/utils/utils.go`. This is solved, can you check and resolve if done properly?

File: framework/tracing/tracer.go [hunk 18]
```diff
@@ -226,256 +226,23 @@ func (t *Tracer) GetDeferredSpanID(traceID string) string {
...
-			ChatNonStreamResponseChoice: &schemas.ChatNonStreamResponseChoice{
-				Message: msg,
-			},
-		}
-		result.ChatResponse.Choices = append(result.ChatResponse.Choices, choice)
-	}
-
-	return result
+	ttftMs, chunkCount := t.store.GetAccumulatedData(traceID)
+	return nil, ttftMs, chunkCount
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply framework/tracing/tracer.go:245]: Fixed — moved TTFT and chunk count attribute setting outside the `accumulatedResp != nil` block in `completeDeferredSpan`, so these metrics are now recorded regardless of whether the accumulated response is nil. This is solved, can you check and resolve if done properly?

File: plugins/logging/operations.go [hunk 19]
```diff
@@ -434,13 +434,203 @@ func (p *LoggerPlugin) updateStreamingLogEntry(
...
+		}
+	}
+}
+
+// applyNonStreamingOutputToEntry applies non-streaming response data to a log entry.
+func (p *LoggerPlugin) applyNonStreamingOutputToEntry(entry *logstore.Log, result *schemas.BifrostResponse) {
+	// Token usage
+	var usage *schemas.BifrostLLMUsage
+	switch {
+	case result.TextCompletionResponse != nil && result.TextCompletionResponse.Usage != nil:
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply plugins/logging/operations.go:544]: Fixed — added `if result == nil { return }` guard at the start of `applyNonStreamingOutputToEntry`. This is solved, can you check and resolve if done properly?

**Comment by akshaydeo** [comment]
@coderabbitai full review

**Comment by akshaydeo** [comment]
@coderabbitai full review

File: transports/bifrost-http/handlers/inference.go [hunk 20]
```diff
@@ -25,6 +25,13 @@ import (
 	"github.com/valyala/fasthttp"
 )
 
+// forwardProviderHeaders forwards provider response headers to the HTTP response.
+func forwardProviderHeaders(ctx *fasthttp.RequestCtx, headers map[string]string) {
+	for key, value := range headers {
+		ctx.Response.Header.Set(key, value)
+	}
+}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply transports/bifrost-http/handlers/inference.go:33]: The filtering is already handled upstream — by the time headers reach `forwardProviderHeaders`, they've already been filtered through `providerResponseFilterHeaders` in `core/providers/utils/utils.go:365-393`.

`ExtractProviderResponseHeaders` / `ExtractProviderResponseHeadersFromHTTP` (which every provider calls when extracting response headers) already excludes all hop-by-hop, framing, and sensitive headers:

- `connection`, `keep-alive`, `transfer-encoding`, `te`, `trailer`, `upgrade`
- `set-cookie`, `set-cookie2`, `cookie`, `authorization`, `www-authenticate`, `proxy-*`
- `content-length`, `content-encoding`, `content-type`
- `host`, `date`, `server`, `alt-svc`, `strict-transport-security`, CORS headers

So `forwardProviderHeaders` at the transport layer only ever receives pre-filtered, safe headers. Adding a second blocklist here would be redundant.

This is solved, can you check and resolve if done properly?

File: plugins/logging/writer.go [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+			if len(batch) >= maxBatchSize {
+				if timerRunning {
+					timer.Stop()
+					timerRunning = false
+				}
+				p.processBatch(batch)
+				batch = batch[:0]
+			} else if !timerRunning {
+				timer.Reset(batchInterval)
+				timerRunning = true
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply plugins/logging/writer.go:72]: Fixed — added the idiomatic non-blocking drain when `timer.Stop()` returns false:

```go
if !timer.Stop() {
    select {
    case <-timer.C:
    default:
    }
}
```

This prevents a stale tick from causing a premature batch flush on the next select iteration.

This is solved, can you check and resolve if done properly?

**Comment by akshaydeo** [comment]
### Merge activity

* **Mar 2, 10:13 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/1868).
